hubblestack/extmods/modules/pulsar.py

```diff
@@ -66,17 +66,529 @@ def _enqueue(revent):
+        # the parent_db sets contain any of the removed dirpaths
+        # and then make sure there's no empty sets in the parent_db
+        to_fully_delete = set()
+        for d,s in salt.ext.six.iteritems(self.parent_db):
+            s -= plist
+            if not s:
+                to_fully_delete.add(d)
+        for item in to_fully_delete:
+            del self.parent_db[item]
+
+    def del_watch(self, wd):
+        ''' remove a watch from the watchmanager database
+        '''
+        if not isinstance(wd, int):
+            wd = self._get_wdl(wd)[0]
+        self.__super.del_watch(wd)
+        self._rm_db(wd)
+
+    def rm_watch(self, *wd, **kw):
+        ''' recursively unwatch things
+        '''
+        wdl = self._get_wdl(wd)
+        res = self.__super.rm_watch(wdl, **kw)
+        self._rm_db(wdl)
+        return res
 
 def _get_notifier():
     '''
     Check the context for the notifier and construct it if not present
     '''
     if 'pulsar.notifier' not in __context__:
         __context__['pulsar.queue'] = collections.deque()
-        wm = pyinotify.WatchManager()
+        log.info("creating new watch manager")
+        wm = PulsarWatchManager()
         __context__['pulsar.notifier'] = pyinotify.Notifier(wm, _enqueue)
     return __context__['pulsar.notifier']
 
+def _preprocess_excludes(excludes):
+    '''
+    Wrap excludes in simple decision curry functions.
+    '''
+
+    # silently discard non-list excludes
+    if not isinstance(excludes, (list,tuple)) or not excludes:
+        return lambda x: False
+
+    # wrap whatever in a decision problem
+    def re_wrapper(robj):
+        # log.debug('wrapping re {0}'.format(robj.pattern))
+        def _wrapped(val):
+            return bool(robj.search(val))
+        return _wrapped
+    def fn_wrapper(rpat):
+        # log.debug('wrapping fnmatch {0}'.format(rpat))
+        def _wrapped(val):
+            return bool(fnmatch.fnmatch(rpat, val))
+        return _wrapped
+    def str_wrapper(rstr):
+        # log.debug('wrapping strmatch {0}'.format(rstr))
+        def _wrapped(val):
+            return bool( val.startswith(rstr) )
+        return _wrapped
+
+    # figure out what to wrap things with
+    the_list = []
+    for e in excludes:
+        if isinstance(e,dict):
+            if e.values()[0].get('regex'):
+                r = e.keys()[0]
+                try:
+                    c = re.compile(r)
+                    the_list.append(re_wrapper(c))
+                except:
+                    log.warn('Failed to compile regex: {0}'.format(r))
+                continue
+            else:
+                e = e.keys()[0]
+        if '*' in e:
+            the_list.append(fn_wrapper(e))
+        else:
+            the_list.append(str_wrapper(e))
+
+    # finally, wrap the whole decision set in a decision wrapper
+    def _final(val):
+        for i in the_list:
+            if i( val ):
+                return True
+        return False
+    time.sleep(2)
```

Comment:
Why?

Reply:
Oh, that's probably a stray debug thing. I thought I got them all ... they were marked with `# debug` so I could find them and get them out.

Reply:
Oh, I should have said also: I intentionally didn't spend much time on docs. It's hard enough to get the docs right when you're sure you're looking at the final-ish form. I wasn't sure how much of this PR would get accepted, so I didn't spend a lot of time on it. I have notes on what to put where though.

Reply:
lol, my test runs are SO FAST without that sleep(2) in there. how did I not notice that was in there all week?